### ISSUE TITLE
[API DESIGN REVIEW] Enable element-wise weighing of outputs in model.fit(..., sample_weight=weights)

### ISSUE THREAD

**Issue by rcasero** [keras-team/keras#10561]
Hi, 

I would like to propose [this patch](https://github.com/rcasero/keras/commit/0d3bfb53215ee7de595d9de2513cda3fa2a927a3) to enable element-wise `sample_weight` in [`model.fit(..., sample_weight=weight)`](https://keras.io/models/model/#fit) so that in addition to e.g.

    w [scalar] * Loss(y_true, y_pred) [2D array]

we can have

    W [2D array] * Loss(y_true, y_pred) [2D array]

(currently, only 1D vectors --one element per training input-- or 2D matrices for temporal series are allowed).

Basically, the patch adds option `sample_weight_mode='element'` to `model.compile()`, and then weights can be passed to `sample_weight` in `model.fit()` as n-D arrays.

Some of the functionality was already there. For example, [`weighted_masked_objective(fn)` in `training_utils.py`](https://github.com/keras-team/keras/blob/454be50414967002197cc40be9d968a16a07f6b9/keras/engine/training_utils.py#L373) contains the code

```python
        score_array = fn(y_true, y_pred)
...
        # apply sample weighting
        if weights is not None:
            # reduce score_array to same ndim as weight array
            ndim = K.ndim(score_array)
            weight_ndim = K.ndim(weights)
            score_array = K.mean(score_array,
                                 axis=list(range(weight_ndim, ndim)))
            score_array *= weights
            score_array /= K.mean(K.cast(K.not_equal(weights, 0), K.floatx()))
```

as if `weights` are the same shape as `score_array`, then `*=` does element-wise multiplication.

I've [forked keras](https://github.com/rcasero/keras) and written a short script to test the path (with `tensorflow-gpu-1.8.0`), testing a small CNN with one or two outputs, and passing `sample_weight_mode` as string, list and dict, and passing `sample_weight` as array, list and dict.

**Note:** New testing script available in comment below. This one remains here for historical reasons:

```python
import os
os.environ['KERAS_BACKEND'] = 'tensorflow'
import keras
import keras.backend as K
import numpy as np

from keras.models import Model, Sequential
from keras.layers import Activation, Conv2D, Input
from keras.layers.normalization import BatchNormalization

# remove warning "Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA"
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '2'  # Just disables the warning, doesn't enable AVX/FMA

K.set_image_data_format('channels_first')

# simulate input images
im = np.zeros(shape=(10, 3, 64, 64), dtype='uint8')

# simulate network output
out = 2 * np.ones(shape=(10, 1, 64, 64), dtype='float32')
aux_out = 5 * np.ones(shape=(10, 1, 22, 22), dtype='float32')
# simulate training weights for network output
weight = np.ones(shape=(10, 1, 64, 64), dtype='float32')
aux_weight = np.ones(shape=(10, 1, 22, 22), dtype='float32')

# simulate validation data
im_validation = 3 * np.ones(shape=(5, 3, 64, 64), dtype='uint8')
out_validation = 4 * np.ones(shape=(5, 1, 64, 64), dtype='float32')
validation_data = (im_validation, out_validation)

# optimizer
optimizer = keras.optimizers.SGD(lr=0.01, decay=1e-6, momentum=0.9, nesterov=True)

'''simple CNN with one output
'''

# create network model
model = Sequential()
model.add(Conv2D(input_shape=(3, 64, 64),
                 filters=32, kernel_size=(3, 3), strides=1, padding='same'))
model.add(BatchNormalization(axis=3))
model.add(Activation('relu'))
model.add(Conv2D(filters=1, kernel_size=(1, 1), strides=1, padding='same'))

'''string format (sample_weights_mode='element')
'''

model.compile(loss='mae', optimizer=optimizer, metrics=['accuracy'], sample_weight_mode='element')

# train with validation_split
##model.fit(im, out, sample_weight=weight, validation_split=0.1, batch_size=3, epochs=3)

# train with validation_data
model.fit(im, out, sample_weight=weight, validation_data=validation_data, batch_size=3, epochs=3)

'''dictionary format (sample_weight_mode={'conv2d_2': 'element'})
'''

model.compile(loss='mae', optimizer=optimizer, metrics=['accuracy'], sample_weight_mode={'conv2d_2': 'element'})

# train with validation_data
model.fit(im, out, sample_weight=weight, validation_data=validation_data, batch_size=3, epochs=3)

'''Multi-output CNN
'''

# create network model
input = Input(shape=(3, 64, 64), dtype='float32')
x = Conv2D(filters=32, kernel_size=(3, 3), strides=1, padding='same')(input)
x = BatchNormalization(axis=3)(x)
x = Activation('relu')(x)

main_output = Conv2D(filters=1, kernel_size=(1, 1), strides=1, padding='same', name='main_output')(x)
aux_output = Conv2D(filters=1, kernel_size=(1, 1), strides=3, padding='same', name='aux_output')(x)

model = Model(inputs=input, outputs=[main_output, aux_output])

'''list format (sample_weight_mode=['element', 'element'])
'''

model.compile(loss='mae', optimizer=optimizer, metrics=['accuracy'],
              sample_weight_mode=['element', 'element'])

model.fit(im, [out, aux_out], sample_weight=[weight, aux_weight],
          batch_size=3, epochs=3)

'''dictionary format (sample_weight_mode={'conv2d_2': 'element'})
'''

model.compile(loss='mae', optimizer=optimizer, metrics=['accuracy'],
              sample_weight_mode={'main_output': 'element', 'aux_output': 'element'})

model.fit(im, {'main_output': out, 'aux_output': aux_out},
          sample_weight={'main_output': weight, 'aux_output': aux_weight},
          batch_size=3, epochs=3)
```

**Comment by rcasero** [comment]
The first change I suggest is adding a new option `sample_weight_mode='element'` to [`model.compile()`](https://keras.io/models/model/#compile). This has to be added to `compile()` in `training.py` in a couple of places, to cover different number of outputs and weight formats (dict, list, array) of `sample_weight_mode`. For example, for the case of a single output and weights given as an array, we'd add to [this `if` block in `training.py`](https://github.com/rcasero/keras/blob/23b5e791adefe4f736f740675c47c4c8f830b505/keras/engine/training.py#L297)

```python
                    if sample_weight_mode == 'temporal':
                        sample_weights.append(
                            K.placeholder(ndim=2,
                                          name=name + '_sample_weights'))
                        sample_weight_modes.append('temporal')
                    else:
                        sample_weights.append(
                            K.placeholder(ndim=1,
                                          name=name + '_sample_weights'))
                        sample_weight_modes.append(None)
```

[this snippet](https://github.com/rcasero/keras/blob/f7f2b5fe837301777df4cc3b55931ef731e01266/keras/engine/training.py#L304)

```python
                    if sample_weight_mode == 'temporal':
                        ...
                    elif sample_weight_mode == 'element':
                        output_shape = self.get_output_shape_at(0)
                        sample_weights.append(
                            K.placeholder(ndim=len(output_shape)-1,
                                          name=name + '_sample_weights'))
                        sample_weight_modes.append('element')
                    else:
                        ...
```

**Comment by rcasero** [comment]
At this point, the test script above runs successfully

```
/home/rcasero/.conda/envs/cytometer_tensorflow/bin/python /snap/pycharm-community/64/helpers/pydev/pydevd.py --multiproc --qt-support=auto --client 127.0.0.1 --port 46211 --file /home/rcasero/Software/cytometer/scripts/keras_branch_sample_weight.py
pydev debugger: process 19509 is connecting

Connected to pydev debugger (build 181.5087.37)
Using TensorFlow backend.
Backend TkAgg is interactive backend. Turning interactive mode on.
Epoch 1/3
Python 3.6.3 |Anaconda, Inc.| (default, Nov  3 2017, 19:19:16) 
[GCC 7.2.0] on linux
[2.0, 1.0]

 3/10 [========>.....................] - ETA: 33s - loss: 1.9810 - acc: 1.0000
10/10 [==============================] - 14s 1s/step - loss: 1.9442 - acc: 1.0000
Epoch 2/3

 3/10 [========>.....................] - ETA: 0s - loss: 1.8317 - acc: 1.0000
10/10 [==============================] - 0s 1ms/step - loss: 1.7663 - acc: 1.0000
Epoch 3/3

 3/10 [========>.....................] - ETA: 0s - loss: 1.5962 - acc: 1.0000
10/10 [==============================] - 0s 1ms/step - loss: 1.5120 - acc: 1.0000

Process finished with exit code 0
```

**Comment by brge17** [comment]
+1

**Comment by rcasero** [comment]
New `sample_weight_mode='element'` was breaking when training with `validation_data` (`validation_split` worked fine). Fixed by [creating a validation weights array of 1.0s with the same size as `val_y` in `model.fit()`](https://github.com/rcasero/keras/commit/e40729eae0609b856a46e69097f03519be72aea2). 

```python
                if self.sample_weight_mode == 'element':
                    val_sample_weight = np.ones(shape=val_y.shape, dtype=float)
                else:
                    val_sample_weight = None
```

A test script for this is

```python
import os
os.environ['KERAS_BACKEND'] = 'tensorflow'
import keras
import keras.backend as K
import numpy as np

from keras.models import Sequential
from keras.layers import Activation, Conv2D
from keras.layers.normalization import BatchNormalization

# remove warning "Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA"
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '2'  # Just disables the warning, doesn't enable AVX/FMA

K.set_image_data_format('channels_first')

# simulate input images
im = np.zeros(shape=(10, 3, 64, 64), dtype='uint8')

# simulate network output
out = 2 * np.ones(shape=(10, 1, 64, 64), dtype='float32')

# simulate training weights for network output
weight = np.ones(shape=(10, 1, 64, 64), dtype='float32')

# simulate validation data
im_validation = 3 * np.ones(shape=(5, 3, 64, 64), dtype='uint8')
out_validation = 4 * np.ones(shape=(5, 1, 64, 64), dtype='float32')
validation_data = (im_validation, out_validation)

# create network model
model = Sequential()
model.add(Conv2D(input_shape=(3, 64, 64),
                 filters=32, kernel_size=(3, 3), strides=1, padding='same'))
model.add(BatchNormalization(axis=3))
model.add(Activation('relu'))
model.add(Conv2D(filters=1, kernel_size=(1, 1), strides=1, padding='same'))

optimizer = keras.optimizers.SGD(lr=0.01, decay=1e-6, momentum=0.9, nesterov=True)

model.compile(loss='mae', optimizer=optimizer, metrics=['accuracy'], sample_weight_mode='element')

# train with validation_split
#model.fit(im, out, sample_weight=weight, validation_split=0.1, batch_size=3, epochs=3)

# train with validation_data
model.fit(im, out, sample_weight=weight, validation_data=validation_data, batch_size=3, epochs=3)
```

**Comment by rcasero** [comment]
Keras API Design Review google doc with comments enabled here

https://docs.google.com/document/d/19BDXgNmeTgpgb9xYKzNboXyM7XX2PeM3mlvCFCdIQj0/edit?usp=sharing

**Comment by rcasero** [comment]
This issue in the keras user mailing list

https://groups.google.com/forum/#!topic/keras-users/IgQYtlSYCp0

**Comment by rcasero** [comment]
Pull request created here

https://github.com/keras-team/keras/pull/10775

as suggested by [David Menéndez Hurtado](https://groups.google.com/d/msg/keras-users/IgQYtlSYCp0/omIa8GazAAAJ).

**Comment by rcasero** [comment]
At this point, some tests fail in the pull request automatic testing. I have asked for guidance in the thread ["How to debug a keras pull request"](https://groups.google.com/forum/#!topic/keras-users/uZo3Z2JkCRA) in the mailing list.

**Comment by rcasero** [comment]
New testing script: 

```python
import os
os.environ['KERAS_BACKEND'] = 'tensorflow'
import keras
import keras.backend as K
import numpy as np

from keras.models import Model, Sequential
from keras.layers import Activation, Conv2D, Input
from keras.layers.normalization import BatchNormalization

from keras.utils import multi_gpu_model

# remove warning "Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA"
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '2'  # Just disables the warning, doesn't enable AVX/FMA

# image_data_format = 'channels_first'
image_data_format = 'channels_last'

K.set_image_data_format(image_data_format)

# simulate input images
if image_data_format == 'channels_first':
    im = np.zeros(shape=(10, 3, 64, 64), dtype='uint8')

    # simulate network output
    out = 2 * np.ones(shape=(10, 1, 64, 64), dtype='float32')
    aux_out = 5 * np.ones(shape=(10, 1, 22, 22), dtype='float32')
    # simulate training weights for network output
    weight = np.ones(shape=(10, 1, 64, 64), dtype='float32')
    aux_weight = np.ones(shape=(10, 1, 22, 22), dtype='float32')

    # simulate validation data
    im_validation = 3 * np.ones(shape=(5, 3, 64, 64), dtype='uint8')
    out_validation = 4 * np.ones(shape=(5, 1, 64, 64), dtype='float32')

elif image_data_format == 'channels_last':
    im = np.zeros(shape=(10, 64, 64, 3), dtype='uint8')

    # simulate network output
    out = 2 * np.ones(shape=(10, 64, 64, 1), dtype='float32')
    aux_out = 5 * np.ones(shape=(10, 22, 22, 1), dtype='float32')
    # simulate training weights for network output
    # weight = np.ones(shape=(10, 64, 64, 1), dtype='float32')
    weight = np.ones(shape=(10, 64, 64, 1), dtype='float32')
    aux_weight = np.ones(shape=(10, 22, 22, 1), dtype='float32')

    # simulate validation data
    im_validation = 3 * np.ones(shape=(5, 64, 64, 3), dtype='uint8')
    out_validation = 4 * np.ones(shape=(5, 64, 64, 1), dtype='float32')

else:
    raise ValueError('Unrecognised position for channels')


validation_data = (im_validation, out_validation)

# optimizer
optimizer = keras.optimizers.SGD(lr=0.01, decay=1e-6, momentum=0.9, nesterov=True)

'''simple CNN with one output
'''

# create network model
model = Sequential()
model.add(Conv2D(input_shape=im.shape[1:],
                 filters=32, kernel_size=(3, 3), strides=1, padding='same'))
model.add(BatchNormalization(axis=3))
model.add(Activation('relu'))
model.add(Conv2D(filters=1, kernel_size=(1, 1), strides=1, padding='same'))

'''string format (sample_weights_mode='element')
'''

# compile model
model.compile(loss='mae', optimizer=optimizer, metrics=['accuracy'], sample_weight_mode='element')

# train with validation_data
model.fit(im, out, sample_weight=weight, validation_data=validation_data, batch_size=3, epochs=3)

'''dictionary format (sample_weight_mode={'conv2d_2': 'element'})
'''

model.compile(loss='mae', optimizer=optimizer, metrics=['accuracy'], sample_weight_mode={'conv2d_2': 'element'})

# train with validation_data
model.fit(im, out, sample_weight=weight, validation_data=validation_data, batch_size=3, epochs=3)


'''parallel training in multiple GPUs
   string format (sample_weights_mode='element')
'''

# list all CPUs and GPUs
device_list = K.get_session().list_devices()

# number of GPUs
gpu_number = np.count_nonzero(['GPU' in str(x) for x in device_list])

if gpu_number > 1:  # compile and train model: Multiple GPUs

    # compile model
    parallel_model = multi_gpu_model(model, gpus=gpu_number)
    parallel_model.compile(loss='mae', optimizer=optimizer, metrics=['accuracy'], sample_weight_mode='element')

    # train model
    parallel_model.fit(im, out, sample_weight=weight, validation_data=validation_data, batch_size=gpu_number, epochs=3)

else:
    raise Warning('Cannot test parallel GPU processing, because I could not find 2 or more GPUs')

'''Multi-output CNN
'''

# create network model
input = Input(shape=im.shape[1:], dtype='float32')
x = Conv2D(filters=32, kernel_size=(3, 3), strides=1, padding='same')(input)
x = BatchNormalization(axis=3)(x)
x = Activation('relu')(x)

main_output = Conv2D(filters=1, kernel_size=(1, 1), strides=1, padding='same', name='main_output')(x)
aux_output = Conv2D(filters=1, kernel_size=(1, 1), strides=3, padding='same', name='aux_output')(x)

model = Model(inputs=input, outputs=[main_output, aux_output])

'''list format (sample_weight_mode=['element', 'element'])
'''

model.compile(loss='mae', optimizer=optimizer, metrics=['accuracy'],
              sample_weight_mode=['element', 'element'])

model.fit(im, [out, aux_out], sample_weight=[weight, aux_weight],
          batch_size=3, epochs=3)

'''dictionary format (sample_weight_mode={'conv2d_2': 'element'})
'''

model.compile(loss='mae', optimizer=optimizer, metrics=['accuracy'],
              sample_weight_mode={'main_output': 'element', 'aux_output': 'element'})

model.fit(im, {'main_output': out, 'aux_output': aux_out},
          sample_weight={'main_output': weight, 'aux_output': aux_weight},
          batch_size=3, epochs=3)
```

**Comment by rcasero** [comment]
[Currently all Travis builds pass](https://travis-ci.org/keras-team/keras/builds/412359858).
